sequence_processing_pipeline/Pipeline.py

```diff
@@ -703,80 +750,72 @@ def _get_sample_names_from_mapping_file(self, project_name):
 
         if short_names is False:
             # return the fully-qualified project name w/Qiita ID.
-            return project_name, matches[2]
+            return project_name, proj_info[QIITA_ID_KEY]
         else:
             # return the project's name and qiita_id
-            return matches[1], matches[2]
+            return proj_info[PROJECT_SHORT_NAME_KEY], proj_info[QIITA_ID_KEY]
 
     def get_project_info(self, short_names=False):
         # test for self.mapping_file, since self.sample_sheet will be
         # defined in both cases.
         results = []
 
+        contains_replicates = None
         if self.mapping_file is not None:
-            if 'contains_replicates' in self.mapping_file:
+            if CONTAINS_REPLICATES_KEY in self.mapping_file:
                 contains_replicates = True
             else:
                 contains_replicates = False
 
             sample_project_map = {pn: _df.sample_name.values for pn, _df in
-                                  self.mapping_file.groupby('project_name')}
-
-            for project in sample_project_map:
-                p_name, q_id = self._parse_project_name(project, short_names)
-                results.append(
-                    {'project_name': p_name, 'qiita_id': q_id,
-                     'contains_replicates': contains_replicates})
+                                  self.mapping_file.groupby(_PROJECT_NAME_KEY)}
+            projects_info = \
+                {p: parse_project_name(p) for p in sample_project_map}
         else:
-            bioinformatics = self.sample_sheet.Bioinformatics
-            for res in bioinformatics.to_dict('records'):
-                p_name, q_id = self._parse_project_name(res['Sample_Project'],
-                                                        short_names)
-
-                # parsed SampleSheet() objects should now contain only
-                # boolean values in contains_replicates column.
-
-                contains_replicates = False
-
-                if 'contains_replicates' in res:
-                    if res['contains_replicates'] is True:
-                        contains_replicates = True
-
-                results.append({'project_name': p_name,
-                                'qiita_id': q_id,
-                                'contains_replicates': contains_replicates})
+            projects_info = self.sample_sheet.get_projects_details()
+        # if this is/isn't a mapping file
+
+        proj_name_key = PROJECT_SHORT_NAME_KEY if short_names \
+            else PROJECT_FULL_NAME_KEY
+        for _, curr_project_info in projects_info.items():
+            curr_dict = {
+                _PROJECT_NAME_KEY: curr_project_info[proj_name_key],
+                QIITA_ID_KEY: curr_project_info[QIITA_ID_KEY]
+            }
+
+            curr_contains_reps = \
+                contains_replicates if contains_replicates is not None else \
+                curr_project_info.get(CONTAINS_REPLICATES_KEY, False)
```

Comment:
Could this be written as an if/else rather than ternary? It's quite complex. I think it also shows that `contain_replicates` can safely default to `False` rather than the `None` currently used

Reply:
I have rewritten this as an if-else.  I do not believe, however, that `contains_replicates` can default to False. 
 The logic is: `contains_replicates` is `True` if there is a mapping file that says it contains replicates, `False` if there is a mapping file that doesn't say it contains replicates, and `None` if there is no mapping file at all.  In that last case, we know we have to get the replicate info for the specific project we are currently looping over, rather than using a universal True or False value.

Reply:
Okay thanks, it looks like it's a ternary still, is the commit coming later? 

Thank's re the default to `False`, I was confused by the original `.get()` :)

I'm not sure I understand the comment about `None`, but let's move on

Reply:
Yeah, sorry, wrote the reply before the push :-/  Here it is:
https://github.com/biocore/mg-scripts/blob/07ff4a3d3a15f9992a96f516eb0f69efce51c44c/sequence_processing_pipeline/Pipeline.py#L808-L814